Feature/nav 113 make raptor use unix timestamps instead of seconds of day

clukas1 opened the pull request.

munterfi commented on src/main/java/ch/naviqore/raptor/Connection.java:
```diff
@@ -1,21 +1,22 @@
     String getToStopId();
 
-    int getDuration();
+    int getDurationInSeconds();
```
Maybe return a `Duration`? https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html

clukas1 replied:
Was thinking of the same but didn't want to over engineer this for now.

munterfi replied:
We could also omit this field completely, since it is kind of redundant?

munterfi commented on src/main/java/ch/naviqore/raptor/router/LabelPostprocessor.java:
```diff
@@ -160,6 +162,9 @@ List<Connection> reconstructParetoOptimalSolutions(List<StopLabelsAndTimes.Label
                 throw new IllegalStateException("Unknown label type");
             }
 
+            LocalDateTime departureTime = LocalDateTime.ofEpochSecond(departureTimestamp, 0, ZoneOffset.UTC);
```
The timezone is defined per agency in the GTFS. So it i not always UTC. But as far as I understand from the docs it is always the same time zone per feed.

> agency_timezone (Timezone | Required): Timezone where the transit agency is located. If multiple agencies are specified in the dataset, each must have the same agency_timezone.

clukas1 replied:
For this application it doesn't matter which Timezone you use in the conversion, as the UTC Timestamp is only handled internally and reconverted to LocalDateTime. The timestamps are never returned outside of the raptor package.

munterfi replied:
Hmm ok, but why then choosing the origin of POSIX time and introduce the need for long? We could just set the minimum time in the builder as origin and stay with ints?

munterfi commented on src/main/java/ch/naviqore/raptor/router/RaptorConnection.java:
```diff
@@ -69,8 +71,8 @@ public String getToStopId() {
-    public int getDuration() {
-        return getArrivalTime() - getDepartureTime();
+    public int getDurationInSeconds() {
+        return (int) Duration.between(getDepartureTime(), getArrivalTime()).getSeconds();
```
This is critical. int could not be enough. There might scenarios where we have timetables afer January 19 , 2038. Which is actually the case in our simulations at work.

Maybe the data source is not GTFS in this cas, but I think since we are building a new routing service, it should be able to handle future timetables.

clukas1 replied:
In this specific case of duration this is not critical unless you want to calculate the duration of a journey that started in 1970 and ended 70 years later. For the date in general I agree, we can convert our unix timestamps to long in this case.

munterfi replied:
Ah sorry, wrong location in the code... im worried about future timetables after 2040...

munterfi reviewed (commented):
Hi @clukas1 

Thanks for the suggestion to move away from seconds per day to unix timestamps.

However, I would vote against this changes, since it introduces a new layer of complexity (handling timzones, future timetables, ...), and I don't see the current benefit. Multi-day routing requests are are more of a rare problem, which could be particularly relevant at weekends. However, night buses should still be included in the timetable for the previous day.

I would rather focus first on, in my opinion more relevant issues:

- Study and implement the interim review feedback.
- Implement the range query in raptor.
- Implement the masking of service trips in raptor to avoid building a new raptor every time. In the current setup this should be quite easy, and i guess when moving to timestamps we have to expand the mask to the complete validity period of the GTFS schedule. I have a bit of a worry this will blow up our memory usage.
- Test, benchmark and document the refactoring (unfortunately we have introduced the same error of not connected legs again `Legs are not connected: RaptorLeg(routeId=transfer_1102155_1102138, tripId=null, fromStopId=1102155, toStopId=1102138, departureTime=24060, arrivalTime=24180, type=WALK_TRANSFER) -> RaptorLeg(routeId=92-A00-3-j24-1_sr1, tripId=328.TA.92-A00-3-j24-1.1.H, fromStopId=1102229, toStopId=1102155, departureTime=24420, arrivalTime=24060, type=ROUTE)`)
- Release a new version of the refactored raptor.

From my point of view, we could move forward towards multi-day queries, when we have solved the issues above.

Nevertheless, I like the change that the inteface of the raptor does not take seconds but LocalDatetimes. However, this can also be implemented without changing the internal structure to UNIX timestamps.

But let's discuss this PR in the meeting just now.

clukas1 commented:
Probably to no surprise I disagree with your opinion to not include timestamps / multi day logic in the raptor. With our current implementation it is not possible to request any connection after midnight which will rely on the calendar services of the previous day. E.g. if I want to get home after midnight in Zürich and request a connection home in Zürich our service will tell us to wait until 5 am for the first tram/bus. The only way you can use a night bus / late connection is by starting before midnight, which makes including night buses etc. in our schedule kind of pointless. This is why we need to have some possibility to handle data of multiple calendar service days at the same time (this can of course also be handled subtracting and adding values by offsets of 24*3600s) but I think timestamps are to be preferred in this case as they are less ambiguous. Also I think it doesn't make sense that the Raptor Connections currently created return int values in seconds of day relative to the start day of the selected raptor instance for the chosen service calendar day, to fix this the raptor instance would have to be aware of it's own date which it currently is not and once again handling timestamps instead of seconds of service calendar day would simplify this.
For the timing of the issue, I think it's essential to implement this as early as possible because it was still simple to implement now and I think we should do this before masking trips etc, because if we don't consider this now it will become significantly more complex at a later stage.
For the masking and memory discussion with timestamps, this is exactly why I wanted to implement this as early as possible to develop a smart solution for this problem. My current vision is to decouple the raptor data structures at least partially from the raptor module and create a raptor data provider interface which can be tweaked based on artifacts of the schedule. So we could more or less use all peculiarities of the GTFS format to store trips with masks and only add the timestamp of the current service day when retrieving the data. Since I'm unaware of how other formats are structured I might be overthinking stuff but I can imagine this way we could keep the memory footprint minimal while still having a lot of flexibility.

clukas1 commented:
> `Legs are not connected: RaptorLeg(routeId=transfer_1102155_1102138, tripId=null, fromStopId=1102155, toStopId=1102138, departureTime=24060, arrivalTime=24180, type=WALK_TRANSFER) -> RaptorLeg(routeId=92-A00-3-j24-1_sr1, tripId=328.TA.92-A00-3-j24-1.1.H, fromStopId=1102229, toStopId=1102155, departureTime=24420, arrivalTime=24060, type=ROUTE))`

Interesting the last time I checked all were working correctly. I've got a couple of tests in mind which I want to write but was waiting for this PR to be closed first, as I don't want to work on the tests of this module on different branches simultaneously.

clukas1 commented:
> > `Legs are not connected: RaptorLeg(routeId=transfer_1102155_1102138, tripId=null, fromStopId=1102155, toStopId=1102138, departureTime=24060, arrivalTime=24180, type=WALK_TRANSFER) -> RaptorLeg(routeId=92-A00-3-j24-1_sr1, tripId=328.TA.92-A00-3-j24-1.1.H, fromStopId=1102229, toStopId=1102155, departureTime=24420, arrivalTime=24060, type=ROUTE))`
> 
> Interesting the last time I checked all were working correctly. I've got a couple of tests in mind which I want to write but was waiting for this PR to be closed first, as I don't want to work on the tests of this module on different branches simultaneously.

@munterfi on which branch did you generate this output? The thing that irritates me about that extract is that it seems that departure time < arrival time in one leg and departure time > arrival time in the other leg. Seems there is some addition/subtraction going wrong in one time type. So it's actually not the same bug but a new one introduced by refactoring.

munterfi commented:
> > > `Legs are not connected: RaptorLeg(routeId=transfer_1102155_1102138, tripId=null, fromStopId=1102155, toStopId=1102138, departureTime=24060, arrivalTime=24180, type=WALK_TRANSFER) -> RaptorLeg(routeId=92-A00-3-j24-1_sr1, tripId=328.TA.92-A00-3-j24-1.1.H, fromStopId=1102229, toStopId=1102155, departureTime=24420, arrivalTime=24060, type=ROUTE))`
> 
> > 
> 
> > Interesting the last time I checked all were working correctly. I've got a couple of tests in mind which I want to write but was waiting for this PR to be closed first, as I don't want to work on the tests of this module on different branches simultaneously.
> 
> 
> 
> @munterfi on which branch did you generate this output? The thing that irritates me about that extract is that it seems that departure time < arrival time in one leg and departure time > arrival time in the other leg. Seems there is some addition/subtraction going wrong in one time type. So it's actually not the same bug but a new one introduced by refactoring.

This was on the current main (before the logging fix).

munterfi commented:
> Probably to no surprise I disagree with your opinion to not include timestamps / multi day logic in the raptor. With our current implementation it is not possible to request any connection after midnight which will rely on the calendar services of the previous day. E.g. if I want to get home after midnight in Zürich and request a connection home in Zürich our service will tell us to wait until 5 am for the first tram/bus. The only way you can use a night bus / late connection is by starting before midnight, which makes including night buses etc. in our schedule kind of pointless. This is why we need to have some possibility to handle data of multiple calendar service days at the same time (this can of course also be handled subtracting and adding values by offsets of 24*3600s) but I think timestamps are to be preferred in this case as they are less ambiguous. Also I think it doesn't make sense that the Raptor Connections currently created return int values in seconds of day relative to the start day of the selected raptor instance for the chosen service calendar day, to fix this the raptor instance would have to be aware of it's own date which it currently is not and once again handling timestamps instead of seconds of service calendar day would simplify this.
> 
> For the timing of the issue, I think it's essential to implement this as early as possible because it was still simple to implement now and I think we should do this before masking trips etc, because if we don't consider this now it will become significantly more complex at a later stage.
> 
> For the masking and memory discussion with timestamps, this is exactly why I wanted to implement this as early as possible to develop a smart solution for this problem. My current vision is to decouple the raptor data structures at least partially from the raptor module and create a raptor data provider interface which can be tweaked based on artifacts of the schedule. So we could more or less use all peculiarities of the GTFS format to store trips with masks and only add the timestamp of the current service day when retrieving the data. Since I'm unaware of how other formats are structured I might be overthinking stuff but I can imagine this way we could keep the memory footprint minimal while still having a lot of flexibility.

Good point, I was not thinking about the issue with departures after midnight.

But still I cannot really see how we implement an efficient trip masking if in the raptor data, all trips are bound to a specific date. We would then still need to shift the dates.

Therefore setting our own artficial origin for the time seems more appropriate to me, and we avoid the need to shift all stoptimes to long.

munterfi commented:
I think we should discuss this issue in a meeting. Maybe we have just different ideas of trip masking, which are currently not compatible 😃

clukas1 commented:
As discussed today, I've removed the to UNIX timestamp conversion and added reference date logic inside raptor and return values as local date time.

To test this extensively, I've also merged NAV-119 into this branch. As mentioned in the other PR that I will now close in favor of this, I've tested it with both Arrival and Departure time types by editing the benchmark script.

Line 128-129:
`LocalDateTime departureTime = SCHEDULE_DATE.atStartOfDay().plusSeconds(random.nextInt(DEPARTURE_TIME_LIMIT));`
becomes:
`LocalDateTime departureTime = SCHEDULE_DATE.atStartOfDay().plusHours(16).plusSeconds(random.nextInt(DEPARTURE_TIME_LIMIT));`

Line 151:
`List<Connection> connections = raptor.routeEarliestArrival(sourceStops, targetStops, new QueryConfig());`
becomes:
`List<Connection> connections = raptor.routeLatestDeparture(targetStops, sourceStops, new QueryConfig());`

munterfi commented on src/main/java/ch/naviqore/raptor/router/DateTimeUtils.java:
```diff
@@ -0,0 +1,48 @@
+        return referenceDate.atStartOfDay().plusSeconds(timestamp);
+    }
+
+    static int convertToTimestamp(LocalDateTime localDateTime, LocalDate referenceDate) {
```
Would this not be convertToSecondsOfDay? Since the timestamp with origin 1970 is removed?

munterfi commented on src/main/java/ch/naviqore/raptor/router/LabelPostprocessor.java:
```diff
@@ -132,19 +136,19 @@ List<Connection> reconstructParetoOptimalSolutions(List<StopLabelsAndTimes.Label
             String toStopId;
-            int departureTime;
-            int arrivalTime;
+            int departureTimestamp;
```
Same issue with naming here?

munterfi commented on src/test/java/ch/naviqore/raptor/router/RaptorRouterTestBuilder.java:
```diff
@@ -79,8 +79,8 @@ private static RaptorAlgorithm build(List<Route> routes, List<Transfer> transfer
             int tripCount = 0;
-            int time = dayStart * SECONDS_IN_HOUR + route.firstDepartureOffset * 60;
-            while (time < dayEnd * SECONDS_IN_HOUR) {
+            int timestamp = dayStart * SECONDS_IN_HOUR + route.firstDepartureOffset * 60;
```
Again, naming...

munterfi reviewed (approved):
Thanks, good to approve from my side.

Just some comments about the naming of "timestamp", but I can live with that.
